File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int f(int i, int amount, vector<int>& coins,vector<vector<int>> &dp){
        if(i == 0){
            if(amount % coins[i] == 0) return amount / coins[i];
            return 1e9;
        }
        
        if(dp[i][amount] != -1) return dp[i][amount];

        int nottake = 0 + f(i-1, amount, coins, dp);
        int take = INT_MAX;
        if(coins[i] <= amount) take = 1 + f(i, amount - coins[i], coins, dp);

        return dp[i][amount] = min(take, nottake);
    }

    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(n, vector<int>(amount + 1, -1));
        int ans = f(n-1, amount, coins, dp);
        if(ans >= 1e9) return -1;
        return ans;
    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def helper(self, ind, coins, amount, dp):
        if ind == 0:
            if amount % coins[ind] == 0:
                return amount // coins[ind]
            return 1e9

        if dp[ind][amount] != -1:
            return dp[ind][amount]

        not_take = self.helper(ind - 1, coins, amount, dp)
        take = 1e9
        if amount >= coins[ind]:
            take = 1 + self.helper(ind, coins, amount - coins[ind], dp)

        dp[ind][amount] = min(take, not_take)
        return dp[ind][amount]

    def coinChange(self, coins: List[int], amount: int) -> int:
        n = len(coins)
        dp = [[-1 for _ in range(amount + 1)] for _ in range(n)]
        ans = self.helper(n-1, coins, amount, dp)
        return -1 if ans >= 1e9 else ans
